keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1368 @@
+and used for multiple different Pods).
+
+Resources are managed by plugins that communicate with central
+Kubernetes components, in particular the kube-scheduler, by updating
+objects stored in the kube-apiserver. Communication with kubelet is
+handled through local Unix domain sockets and the plugin registration
+mechanism, using a new plugin type and a new gRPC interface. Towards
+the container runtime, kubelet and those plugins then must use the
+[Container Device Interface
+(CDI)](https://github.com/container-orchestrated-devices/container-device-interface)
+to expose the resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources are too limited:
+
+- *Device initialization*: When starting a workload I’d like to have
+  the device reconfigured or reprogrammed by the orchestration. For
+  security reasons workloads should not be able to reconfigure devices
+  directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device.
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
+  the node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
```

Comment:
This is a good point!  I think addressing this limitation will help Kubernetes to support more use cases

Reply:
It is fine that allocation fails on the time of pod start time if the request for resource is optional. If we are talking about ReplicaSet of similar deployments with dynamic changing set of running pods such optional allocation failures is expected and valid behaviour.

Absence of requested class is definitely an error.

Reply:
> It is fine that allocation fails on the time of pod start time if the request for resource is optional.

Allocation as designed at the moment happens before the pod is allowed to start because information about where the resource was allocated is needed for choose the node for pod.

Are you suggesting that for optional resources, pod scheduling should proceed without waiting for the resource allocation and then for the assigned node, allocation is tried once with "resource not available" not blocking pod startup?

Reply:
No, what I'm suggesting, is that in case of "optional", allocation by CDI plugin will succeed, returning topology key for any node but actual "allocation" would be "nil". meaning, CDI json might be empty - no new devices injected into OCI spec.

Reply:
So that's the "can be covered by a plugin specific parameter" approach that I mentioned earlier. I'm fine with that, in particular because it keeps the KEP simpler. If we need something more complex, we can still extend the KEP later.

I'll add a section explaining this approach.